File: lib/python/opentoken/src/main/opentoken/attributes/general/date_attribute.py
"""
Copyright (c) Truveta. All rights reserved.
"""

from typing import List
import re
from datetime import datetime
from opentoken.attributes.base_attribute import BaseAttribute
from opentoken.attributes.validation import RegexValidator
from opentoken.attributes.validation.serializable_attribute_validator import SerializableAttributeValidator


class DateAttribute(BaseAttribute):
    """Represents a generic date attribute.

    This class extends BaseAttribute and provides functionality for working with
    date fields. It recognizes "Date" as a valid alias for this attribute type.

    The attribute performs normalization on input values, converting them to a
    standard format (yyyy-MM-dd).

    Supported formats:
    - yyyy-MM-dd
    - yyyy/MM/dd
    - MM/dd/yyyy
    - MM-dd-yyyy
    - dd.MM.yyyy
    """

    NAME = "Date"
    ALIASES = [NAME]

    # Regular expression pattern for validating date formats
    VALIDATION_PATTERN = re.compile(
        r"^(?:\d{4}[-/]\d{2}[-/]\d{2}|\d{2}[-./]\d{2}[-./]\d{4})$"
    )

    def __init__(self, additional_validators: List[SerializableAttributeValidator] = None):
        """
        Initialize the DateAttribute with optional additional validators.

        Args:
            additional_validators: Optional list of additional validators to apply
        """
        validation_rules = [RegexValidator(self.VALIDATION_PATTERN)]
        if additional_validators:
            validation_rules.extend(additional_validators)
        super().__init__(validation_rules)

    def get_name(self) -> str:
        """Get the name of the attribute.

        Returns:
            str: The name "Date"
        """
        return self.NAME

    def get_aliases(self) -> List[str]:
        """Get the aliases for the attribute.

        Returns:
            List[str]: A list containing the aliases for this attribute
        """
        return self.ALIASES.copy()

    def normalize(self, value: str) -> str:
        """Normalizes date to yyyy-MM-dd format.

        Args:
            value: The date string to normalize

        Returns:
            str: The normalized date in yyyy-MM-dd format

        Raises:
            ValueError: If the date format is invalid
        """
        value = value.strip()

        if not value:
            raise ValueError("Invalid date format: empty or whitespace")

        # Try different date formats
        date_formats = [
            "%Y-%m-%d",  # yyyy-MM-dd
            "%Y/%m/%d",  # yyyy/MM/dd
            "%m/%d/%Y",  # MM/dd/yyyy
            "%m-%d-%Y",  # MM-dd-yyyy
            "%d.%m.%Y"   # dd.MM.yyyy
        ]

        for format in date_formats:
            try:
                parsed_date = datetime.strptime(value, format)
                return parsed_date.strftime("%Y-%m-%d")
            except ValueError:
                continue

        raise ValueError(f"Invalid date format: {value}")

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/attributes/general/DateAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes.general;

import java.text.ParseException;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import org.apache.commons.lang3.time.DateUtils;

import com.truveta.opentoken.attributes.BaseAttribute;
import com.truveta.opentoken.attributes.validation.RegexValidator;
import com.truveta.opentoken.attributes.validation.SerializableAttributeValidator;

/**
 * Represents a generic date attribute.
 * 
 * This class extends BaseAttribute and provides functionality for working with
 * date fields. It recognizes "Date" as the primary alias for this attribute type.
 * 
 * The attribute performs normalization on input values, converting them to a
 * standard format (yyyy-MM-dd).
 * 
 * The attribute also performs validation on input values, ensuring they match
 * one of the following formats:
 * - yyyy-MM-dd
 * - yyyy/MM/dd
 * - MM/dd/yyyy
 * - MM-dd-yyyy
 * - dd.MM.yyyy
 */
public class DateAttribute extends BaseAttribute {

    private static final String NAME = "Date";
    private static final String[] ALIASES = new String[] { NAME };

    /**
     * Regular expression pattern for validating date formats.
     * 
     * This regex supports two formats:
     * 1. "YYYY-MM-DD" or "YYYY/MM/DD" - where the year is represented by 4 digits,
     *    followed by a hyphen or slash, then a 2-digit month, another hyphen or slash,
     *    and finally a 2-digit day.
     * 2. "MM-DD-YYYY" or "MM.DD.YYYY" or "MM/DD/YYYY" - where the month is represented
     *    by 2 digits, followed by a hyphen, dot, or slash, then a 2-digit day, and
     *    finally a 4-digit year.
     * 
     * This ensures that the input matches common date formats while allowing for
     * flexibility in the delimiter used (hyphen, slash, or dot).
     */
    private static final String DATE_REGEX = "^((\\d{4}[-/]\\d{2}[-/]\\d{2})|(\\d{2}[-/.]\\d{2}[-/.]\\d{4}))$";

    private static final String NORMALIZED_FORMAT = "yyyy-MM-dd";
    private static final String[] POSSIBLE_INPUT_FORMATS = new String[] {
            NORMALIZED_FORMAT, "yyyy/MM/dd", "MM/dd/yyyy",
            "MM-dd-yyyy", "dd.MM.yyyy" };

    // Thread-safe date formatter
    private static final DateTimeFormatter NORMALIZED_DATE_FORMATTER = DateTimeFormatter.ofPattern(NORMALIZED_FORMAT);

    public DateAttribute() {
        super(List.of(new RegexValidator(DATE_REGEX)));
    }

    /**
     * Protected constructor for subclasses to add additional validators.
     * 
     * @param additionalValidators additional validators to apply beyond the base date format validation
     */
    protected DateAttribute(List<SerializableAttributeValidator> additionalValidators) {
        super(createValidatorList(additionalValidators));
    }

    private static List<SerializableAttributeValidator> createValidatorList(
            List<SerializableAttributeValidator> additionalValidators) {
        java.util.ArrayList<SerializableAttributeValidator> validators = new java.util.ArrayList<>();
        validators.add(new RegexValidator(DATE_REGEX));
        validators.addAll(additionalValidators);
        return validators;
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public String[] getAliases() {
        return ALIASES;
    }

    @Override
    public String normalize(String value) {
        try {
            // We still need to use DateUtils to parse the various formats
            Date date = DateUtils.parseDateStrictly(value, Locale.ENGLISH, POSSIBLE_INPUT_FORMATS);

            // Convert Date to LocalDate and format using thread-safe DateTimeFormatter
            LocalDate localDate = date.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();

            // Format the LocalDate to the normalized format
            return NORMALIZED_DATE_FORMATTER.format(localDate);
        } catch (ParseException e) {
            throw new IllegalArgumentException("Invalid date format: " + value);
        }
    }

}
